cityofaustin/atd-moped
# Refactor filterParameters state to use search params state shape and filter config

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/15049
https://github.com/cityofaustin/atd-data-tech/issues/15074

This PR:
- updates the `filterParameters` state to have the same shape and key-value pairs (field, operator, and value) as the `filters` state that powers the query
- moves a lot of functions defined within the `Filters` component into a helpers file
- removes code that used to access configuration about a field and then store it in state so that it now accessing that configuration as needed

If you want to visualize the biggest change, run locally and log `filterParameters` in the `Filters` component as you add advanced filters. Compare what that looks like with this branch versus `main` branch.

## Testing
**URL to test:** 

https://deploy-preview-1236--atd-moped-main.netlify.app/moped/projects

**Steps to test:**
1. Go to the project list view and add advanced filters
2. Test a filter with lookup tables like **Type**
3. Test editing applied filters options (choose new field option, choose new operator option, or update search value)
4. Test deleting filters with the trash can icon. A new search should not happen until you click the Search button.
5. Test clearing the first **Field** input after selecting a value by using the X icon that appears to the left of the dropdown icon. Both the **Field** and **Operator** values should clear from the UI.
6. Test resetting the filters by clicking the reset button
7. Test making edits to the url string to make sure the app doesn't break when it can't parse the URL params
8. Test filtering, clicking on a project name link in the list view, and then clicking the "< ALL PROJECTS" link to go back to the list view. Your filters should still be present.
9. Add a filter that expects a number value like **ID is** and try entering letters. You should see that the recently-added validation still works.
10. Check that the trash can delete icons and match any/all radios do not appear until more than one filter is added.

---
#### Ship list
